src/selectors/columnWidths.js

```diff
@@ -0,0 +1,128 @@
+/**
+ * Copyright Schrodinger, LLC
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule columnWidths
+ */
+import { getTotalFlexGrow, getTotalWidth } from 'widthHelper';
+import Scrollbar from 'Scrollbar';
+import forEach from 'lodash/forEach';
+import map from 'lodash/map';
+import partition from 'lodash/partition';
+import scrollbarsVisible from 'scrollbarsVisible';
+import shallowEqualSelector from 'shallowEqualSelector';
+
+/**
+ * @typedef {{
+ *   fixed: boolean,
+ *   flexGrow: number,
+ *   width: number,
+ * }}
+ */
+let columnDefinition;
+
+/**
+ * @param {!Array.<columnDefinition>} columnGroupProps
+ * @param {!Array.<columnDefinition>} columnProps
+ * @param {boolean} scrollEnabledY
+ * @param {number} width
+ * @return {{
+ *   columnGroupProps: !Array.<columnDefinition>,
+ *   columnProps: !Array.<columnDefinition>,
+ *   availableScrollWidth: number,
+ *   fixedColumns: !Array.<columnDefinition>,
+ *   scrollableColumns: !Array.<columnDefinition>,
+ *   maxScrollX: number,
+ * }} The total width of all columns.
+ */
+function columnWidths(columnGroupProps, columnProps, scrollEnabledY, width) {
+  const scrollbarSpace = scrollEnabledY ? Scrollbar.SIZE + Scrollbar.OFFSET : 0;
+  const viewportWidth = width - scrollbarSpace;
+
+  const {
+    newColumnGroupProps,
+    newColumnProps,
+  } = flexWidths(columnGroupProps, columnProps, viewportWidth);
+  const [
+    fixedColumns,
+    scrollableColumns,
+  ] = partition(newColumnProps, column => column.fixed);
+
+  const availableScrollWidth = viewportWidth - getTotalWidth(fixedColumns);
+  const maxScrollX = Math.max(0, getTotalWidth(newColumnProps) - viewportWidth);
+  return {
+    columnGroupProps: newColumnGroupProps,
+    columnProps: newColumnProps,
+    availableScrollWidth,
+    fixedColumns,
+    scrollableColumns,
+    maxScrollX,
+  };
+}
+
+/**
+ * @param {!Array.<columnDefinition>} columnGroupProps
+ * @param {!Array.<columnDefinition>} columnProps
+ * @param {number} viewportWidth
+ * @return {{
+ *   newColumnGroupProps: !Array.<columnDefinition>,
+ *   newColumnProps: !Array.<columnDefinition>
+ * }}
+ */
+function flexWidths(columnGroupProps, columnProps, viewportWidth) {
```

Comment:
new algorithm for expanding flex widths, takes a more linear approach than what was previously in WidthHelper.